File: backend/src/test/java/com/familyledger/MoneyTest.java
package com.familyledger;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.familyledger.common.Money;
import org.junit.jupiter.api.Test;

class MoneyTest {

  @Test
  void toCents_字符串与数字元转分() {
    assertThat(Money.toCents("12.34")).isEqualTo(1234L);
    assertThat(Money.toCents(12.5)).isEqualTo(1250L);
    assertThat(Money.toCents(99.99)).isEqualTo(9999L);
    assertThat(Money.toCents("100")).isEqualTo(10000L);
    assertThat(Money.toCents(100)).isEqualTo(10000L);
    assertThat(Money.toCents("0.01")).isEqualTo(1L);
    assertThat(Money.toCents("-0.05")).isEqualTo(-5L);
    assertThat(Money.toCents("0")).isEqualTo(0L);
  }

  @Test
  void toCents_第三位小数四舍五入到分() {
    assertThat(Money.toCents("1.005")).isEqualTo(101L); // 100 + 进位 1
    assertThat(Money.toCents("1.004")).isEqualTo(100L); // 100，不舍入
    assertThat(Money.toCents("1.999")).isEqualTo(200L); // 199.9 -> 200
  }

  @Test
  void toCents_非法输入抛异常() {
    assertThatThrownBy(() -> Money.toCents("abc")).isInstanceOf(IllegalArgumentException.class);
    assertThatThrownBy(() -> Money.toCents("")).isInstanceOf(IllegalArgumentException.class);
    assertThatThrownBy(() -> Money.toCents(null)).isInstanceOf(IllegalArgumentException.class);
    assertThatThrownBy(() -> Money.toCents("12.34.5")).isInstanceOf(IllegalArgumentException.class);
  }

  @Test
  void toYuanString_分转两位元() {
    assertThat(Money.toYuanString(1234L)).isEqualTo("12.34");
    assertThat(Money.toYuanString(-5L)).isEqualTo("-0.05");
    assertThat(Money.toYuanString(0L)).isEqualTo("0.00");
    assertThat(Money.toYuanString(100000000L)).isEqualTo("1000000.00");
    assertThat(Money.toYuanString(-13024089L)).isEqualTo("-130240.89");
  }
}

File: server/src/domain/money.test.ts
import { describe, it, expect } from 'vitest';
import { toCents, toYuanString } from './money';

describe('toCents', () => {
  it('四舍五入规避浮点误差：12.345 -> 1235', () => {
    expect(toCents('12.345')).toBe(1235);
  });

  it('0.1 -> 10（浮点 0.1*100 不丢精度）', () => {
    expect(toCents(0.1)).toBe(10);
  });

  it('0.1 + 0.2 这条路精确 -> 30', () => {
    expect(toCents(0.1 + 0.2)).toBe(30);
  });

  it('普通两位小数 12.34 -> 1234', () => {
    expect(toCents('12.34')).toBe(1234);
  });

  it('超过两位的舍入 12.3456 -> 1235', () => {
    expect(toCents('12.3456')).toBe(1235);
  });

  it('允许负数 -0.05 -> -5', () => {
    expect(toCents('-0.05')).toBe(-5);
  });

  it('整数 0 -> 0', () => {
    expect(toCents('0')).toBe(0);
  });

  it('空字符串抛错', () => {
    expect(() => toCents('')).toThrow();
  });

  it('非数字字符串抛错', () => {
    expect(() => toCents('abc')).toThrow();
  });

  it('NaN 抛错', () => {
    expect(() => toCents(NaN)).toThrow();
  });
});

describe('toYuanString', () => {
  it('-5 -> "-0.05"', () => {
    expect(toYuanString(-5)).toBe('-0.05');
  });

  it('1234 -> "12.34"', () => {
    expect(toYuanString(1234)).toBe('12.34');
  });

  it('0 -> "0.00"', () => {
    expect(toYuanString(0)).toBe('0.00');
  });

  it('5 -> "0.05"', () => {
    expect(toYuanString(5)).toBe('0.05');
  });

  it('100 -> "1.00"', () => {
    expect(toYuanString(100)).toBe('1.00');
  });

  it('-1234 -> "-12.34"', () => {
    expect(toYuanString(-1234)).toBe('-12.34');
  });
});
